### PR TITLE
fix/issue-1723/nested-reference-queries

### PR DESCRIPTION
**Pull request by TiMESPLiNTER**
See issue: https://github.com/doctrine/mongodb-odm/issues/1723

### CODE REVIEW

File: tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php [hunk 1]
```diff
@@ -359,6 +361,37 @@ public function testQueryReferences()
...
+        $this->dm->persist($user);
+        $this->dm->flush();
+
+        $qb = $this->dm->createQueryBuilder('Documents\ReferenceUser');
+
+        $referencedUsersQuery = $qb
+            ->field('indirectlyReferencedUsers.user.id')->equals($referencedUser->getId())
+            ->getQuery();
+
+        print_r($referencedUsersQuery->getQuery());
```
Comment (malarzm) [inline tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php]: This should be removed

File: tests/Documents/ReferenceUser.php [hunk 2]
```diff
@@ -72,6 +72,13 @@ class ReferenceUser
      */
     public $referencedUsers = array();
 
+    /**
+     * @ODM\EmbedMany(targetDocument="Documents\IndirectlyReferencedUser")
+     *
+     * @var IndirectlyReferencedUser[]
+     */
+    public $indirectlyReferencedUsers = array();
```
Comment (malarzm) [inline tests/Documents/ReferenceUser.php]: Please use short array syntax

**Comment by malarzm** [comment]
@TiMESPLiNTER do you also want to tackle the actual issue or shall I? :)

**Comment by malarzm** [comment]
All right, with this patch something is found:

```
All right, with this patch results are found:

```
--- a/lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php
+++ b/lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php
@@ -1158,7 +1158,7 @@ class DocumentPersister
         }
 
         // No further processing for fields without a targetDocument mapping
-        if ( ! isset($mapping['targetDocument'])) {
+        if ( ! isset($mapping['targetDocument']) || ! empty($mapping['embedded'])) {
             if ($nextObjectProperty) {
                 $fieldName .= '.'.$nextObjectProperty;
             }
diff --git a/tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php b/tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php
index b69a474..54fee95 100644
--- a/tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php
+++ b/tests/Doctrine/ODM/MongoDB/Tests/Functional/QueryTest.php
@@ -381,12 +381,12 @@ class QueryTest extends \Doctrine\ODM\MongoDB\Tests\BaseTest
         $qb = $this->dm->createQueryBuilder('Documents\ReferenceUser');
 
         $referencedUsersQuery = $qb
-            ->field('indirectlyReferencedUsers.user.id')->equals($referencedUser->getId())
+            ->field('indirectlyReferencedUsers.user.id')->equals(new \MongoDB\BSON\ObjectId($referencedUser->getId()))
             ->getQuery();
 
         print_r($referencedUsersQuery->getQuery());
 
-        $referencedUsers = $referencedUsersQuery->execute();
+        $referencedUsers = iterator_to_array($referencedUsersQuery->execute());
 
         $this->assertCount(1, $referencedUsers);
         $this->assertSame($referencedUser, $referencedUsers[0]);
```

So now to why:

`if ( ! isset($mapping['targetDocument']) || ! empty($mapping['embedded'])) {` line is IMO correct fix as for embedded documents we still want further logic to apply, it's just the inner-parts of references we don't want to touch.

`->field('indirectlyReferencedUsers.user.id')->equals(new \MongoDB\BSON\ObjectId($referencedUser->getId()))` - as builder will not prepare the value we must ourselves tell ODM that the id should be `ObjectId`

The test is still failing as last assertion can't be true as `$referencedUser` is of type `User` while you're looking up documents of type `ReferenceUser`. Prob this should check with `$user`?

**Comment by TiMESPLiNTER** [comment]
@malarzm cool, thanks for the patch. I applied it and fixed all the other stuff you mentioned.

Unfortunately now quite a few other test cases break from the test suite. They don't seem to like that change. (Weird results.)

**Comment by malarzm** [comment]
Shoot, I should have ran whole test suite :) I'll see what I can do, but not sure if today

**Comment by alcaeus** [comment]
Pushing back to 1.2.3 milestone - @malarzm please let me know once you've got a chance to take a look.

**Comment by devorphea** [comment]
Hi guys, 
Do you think this issue will be merged soon?

**Comment by malarzm** [comment]
Sorry, it fell through the cracks. I'll be offline for couple of days and middle of next week is earliest I can devote some time to writing code.

**Comment by devorphea** [comment]
Ok great, i hope you can do something for us at the end of next week.

**Comment by devorphea** [comment]
It seems you didn't have time to look at this fix.
Can we do something for helping to merge this fix?

**Comment by malarzm** [comment]
Superseded by #1793, @TiMESPLiNTER I've cherry-picked your commit with test there - thank you very much for writing it!

**Pull request closed without merging**